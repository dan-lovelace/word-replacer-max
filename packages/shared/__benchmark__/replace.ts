import Benchmark from "benchmark";

import { regExpReplace } from "@worm/shared/src/replace/lib/regex";

const suite = new Benchmark.Suite();

suite.add("standard replace", function () {
  "hello world".replace(/(hello)\s(world)/, "$2 $1");
});

suite.add("custom regexp replace", function () {
  regExpReplace("\\U$2\\E \\L$1\\E")("", "hello", "world");
});

suite
  .on("cycle", function (event: Event) {
    console.log(String(event.target));
  })
  .on("complete", function () {
    console.log("Fastest is " + this.filter("fastest").map("name"));
  })
  .run({ async: true });
