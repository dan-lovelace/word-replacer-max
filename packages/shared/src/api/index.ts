import { ApiRouteIdentifier } from "@worm/types/src/api";

import { getEnvConfig } from "../config";

import { getApiRoute } from "./endpoints";

export function getApiEndpoint(name: ApiRouteIdentifier) {
  const envConfig = getEnvConfig();

  return `${envConfig.VITE_API_ORIGIN}${getApiRoute(name)}`;
}
