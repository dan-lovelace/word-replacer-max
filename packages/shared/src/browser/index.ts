export * from "./assets";
export * from "./browser";
export * from "./matchers";
export * from "./popup";
