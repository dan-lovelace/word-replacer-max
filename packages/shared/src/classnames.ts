export function cx(...classes: any[]) {
  return classes
    .filter(Boolean)
    .map((s) => String(s).trim())
    .join(" ");
}
