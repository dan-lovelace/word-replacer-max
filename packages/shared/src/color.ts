class ColorGenerator {
  private recentHues: number[] = [];
  private readonly HUE_DIFFERENCE_THRESHOLD = 30;
  private readonly SATURATION_RANGE = { min: 70, max: 80 };
  private readonly LIGHTNESS_RANGE = { min: 45, max: 65 };
  private readonly UNIQUE_COLOR_MEMORY_SIZE = 5; // Number of previous colors to ensure uniqueness against

  private hslToHex(hue: number, saturation: number, lightness: number): string {
    const hslToRgb = (
      chromaMax: number,
      chromaMin: number,
      huePrime: number
    ): number => {
      if (huePrime < 0) huePrime += 1;
      if (huePrime > 1) huePrime -= 1;
      if (huePrime < 1 / 6)
        return chromaMax + (chromaMin - chromaMax) * 6 * huePrime;
      if (huePrime < 1 / 2) return chromaMin;
      if (huePrime < 2 / 3)
        return chromaMax + (chromaMin - chromaMax) * (2 / 3 - huePrime) * 6;

      return chromaMax;
    };

    saturation /= 100;
    lightness /= 100;

    // Calculate chroma (color intensity) and grayscale adjustment
    const chromaMax =
      lightness < 0.5
        ? lightness * (1 + saturation)
        : lightness + saturation - lightness * saturation;
    const chromaMin = 2 * lightness - chromaMax;

    const r = Math.round(
      hslToRgb(chromaMin, chromaMax, hue / 360 + 1 / 3) * 255
    );
    const g = Math.round(hslToRgb(chromaMin, chromaMax, hue / 360) * 255);
    const b = Math.round(
      hslToRgb(chromaMin, chromaMax, hue / 360 - 1 / 3) * 255
    );

    return "#" + [r, g, b].map((x) => x.toString(16).padStart(2, "0")).join("");
  }

  private isTooSimilar(hue: number): boolean {
    return this.recentHues.some((recentHue) => {
      const diff = Math.abs(hue - recentHue);

      return Math.min(diff, 360 - diff) < this.HUE_DIFFERENCE_THRESHOLD;
    });
  }

  private randomInRange(min: number, max: number): number {
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }

  private generateUniqueHue(): number {
    let hue: number;
    let attempts = 0;
    const MAX_ATTEMPTS = 100;

    do {
      hue = this.randomInRange(0, 359);
      attempts++;

      if (attempts >= MAX_ATTEMPTS) {
        this.recentHues = []; // Reset if we can't find a unique hue

        break;
      }
    } while (this.isTooSimilar(hue));

    this.recentHues.push(hue);

    if (this.recentHues.length > this.UNIQUE_COLOR_MEMORY_SIZE) {
      this.recentHues.shift();
    }

    return hue;
  }

  /**
   * Generates a new semi-random HEX color (i.e. `#ff11cd`). Uniqueness is
   * enforced by keeping a history of generated colors and ensuring any new
   * colors are not too-similar to any in memory.
   */
  public generate(): string {
    const hue = this.generateUniqueHue();
    const saturation = this.randomInRange(
      this.SATURATION_RANGE.min,
      this.SATURATION_RANGE.max
    );
    const lightness = this.randomInRange(
      this.LIGHTNESS_RANGE.min,
      this.LIGHTNESS_RANGE.max
    );

    return this.hslToHex(hue, saturation, lightness);
  }
}

export const colorGenerator = new ColorGenerator();
