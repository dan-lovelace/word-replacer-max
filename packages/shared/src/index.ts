import "../vite-env.d.ts";

export * from "./classnames";
export * from "./debounce";
export * from "./domains";
export * from "./logging";
export * from "./messaging";
export * from "./objects";
export * from "./replace";
export * from "./schemas";
export * from "./selectors";
export * from "./strings";
export * from "./time";
