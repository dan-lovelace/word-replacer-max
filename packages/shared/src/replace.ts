import { Matcher, QueryPattern } from "@worm/types";

import { logDebug } from "./logging";

const escapeRegex = (str: string) =>
  str.replace(/[/\-\\^$*+?.()|[\]{}]/g, "\\$&");

const parentNodeBlocklist: Node["nodeName"][] = [
  "i",
  "img",
  "link",
  "meta",
  "script",
  "style",
  "svg",
  "video",
];

const patternRegex: {
  [key in QueryPattern]: (query: string, flags?: string) => RegExp;
} = {
  case: (query) => new RegExp(query, "g"),
  default: (query) => new RegExp(query, "gi"),
  regex: (query, flags) => new RegExp(query, flags),
  wholeWord: (query, flags) =>
    new RegExp(`(^|\\s)${escapeRegex(query)}($|\\s)`, flags),
};

function getFlags(queryPatterns: QueryPattern[]) {
  let flags = "g";

  if (!queryPatterns.includes("case")) {
    flags += "i";
  }

  return flags;
}

export function searchAndReplace(
  element: HTMLElement,
  query: string,
  { queryPatterns, replacement }: Matcher
) {
  const searchResults = searchNode(element, query, queryPatterns) || [];

  for (const result of searchResults) {
    replace(result, query, queryPatterns, replacement);
  }
}

export function searchNode(
  element: HTMLElement,
  query: string,
  queryPatterns: QueryPattern[],
  found: HTMLElement[] = []
) {
  const elementContents = String(element.textContent);
  let containsText = false;

  if (!queryPatterns || queryPatterns.length < 1) {
    // default query pattern
    containsText = patternRegex.default(query).test(elementContents);
  } else {
    for (const pattern of queryPatterns) {
      switch (pattern) {
        case "case":
        case "default": {
          containsText = patternRegex[pattern](query).test(elementContents);
          break;
        }

        case "regex":
        case "wholeWord": {
          containsText = patternRegex[pattern](
            query,
            getFlags(queryPatterns)
          ).test(elementContents);
          break;
        }
      }
    }
  }

  if (!containsText) {
    return found;
  }

  if (element.hasChildNodes()) {
    const childElements = Array.from(element.childNodes) as HTMLElement[];

    for (const child of childElements) {
      searchNode(child, query, queryPatterns, found);
    }
  }

  if (
    element.nodeType === Node.TEXT_NODE &&
    !parentNodeBlocklist.includes(
      String(element.parentNode?.nodeName.toLowerCase())
    )
  ) {
    found.push(element);
  }

  return found;
}

export function replace(
  element: HTMLElement | null,
  query: string,
  queryPatterns: QueryPattern[],
  replacement: string
) {
  if (element === null) return;

  const { parentElement } = element;

  if (parentElement?.dataset["isReplaced"]) {
    // already replaced
    return;
  }

  const contentsAttribute: keyof HTMLElement = "textContent";
  const elementContents = String(element[contentsAttribute]);

  if (!queryPatterns || queryPatterns.length < 1) {
    // default query pattern
    element[contentsAttribute] = elementContents.replace(
      patternRegex.default(query),
      replacement
    );
  } else {
    for (const pattern of queryPatterns) {
      switch (pattern) {
        case "case":
        case "default": {
          element[contentsAttribute] = elementContents.replace(
            patternRegex[pattern](query),
            replacement
          );
          break;
        }

        case "regex": {
          element[contentsAttribute] = elementContents.replace(
            patternRegex[pattern](query, getFlags(queryPatterns)),
            replacement
          );
          break;
        }

        case "wholeWord": {
          element[contentsAttribute] = elementContents
            .replace(
              patternRegex[pattern](query, getFlags(queryPatterns)),
              ` ${replacement} `
            )
            .replace(/\s\s+/g, "")
            .trim();
          break;
        }
      }
    }
  }

  if (parentElement) {
    parentElement.dataset["isReplaced"] = new Date().getTime().toString();
  }
}

export function replaceAll(matchers: Matcher[], htmlStart?: HTMLHtmlElement) {
  const startAtElement = htmlStart ?? document;

  const body = startAtElement.querySelector("body");
  if (!body) {
    logDebug("No `body` element found");
  }

  const head = startAtElement.querySelector("head");
  if (!head) {
    logDebug("No `head` element found");
  }

  for (const matcher of matchers) {
    if (matcher.active !== true) continue;

    for (const query of matcher.queries) {
      if (body) {
        searchAndReplace(body, query, matcher);
      }

      if (head) {
        for (const title of Array.from(head.querySelectorAll("title"))) {
          searchAndReplace(title, query, matcher);
        }
      }
    }
  }
}
