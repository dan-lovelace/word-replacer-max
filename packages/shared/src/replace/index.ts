import { Matcher } from "@worm/types";

import { findText } from "./find-text";
import { replaceText } from "./replace-text";

import { logDebug } from "../logging";

function searchAndReplace(
  element: HTMLElement,
  query: string,
  { queryPatterns, replacement }: Matcher
) {
  const searchResults = findText(element, query, queryPatterns) || [];

  for (let position = 0; position < searchResults.length; position++) {
    const result = searchResults[position];

    replaceText(result, query, queryPatterns, replacement, position);
  }
}

export function replaceAll(matchers: Matcher[], startDocument = document) {
  const body = startDocument.querySelector("body");
  if (!body) {
    logDebug("No `body` element found");
  }

  const head = startDocument.querySelector("head");
  if (!head) {
    logDebug("No `head` element found");
  }

  for (const matcher of matchers) {
    if (matcher.active !== true) continue;

    for (const query of matcher.queries) {
      if (body) {
        searchAndReplace(body, query, matcher);
      }

      if (head) {
        for (const title of Array.from(head.querySelectorAll("title"))) {
          searchAndReplace(title, query, matcher);
        }
      }
    }
  }
}

export * from "./find-text";
export * from "./lib";
export * from "./replace-text";
