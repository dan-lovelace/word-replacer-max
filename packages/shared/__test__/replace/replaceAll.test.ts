import { replaceAll } from "@worm/shared/src/replace";

jest.useFakeTimers().setSystemTime(new Date("2024-01-01"));

describe("replaceAll", () => {
  // describe("document cases", () => {
  //   it("includes 'title' elements in the head", () => {
  //     document.documentElement.innerHTML = `
  //       <html>
  //         <head>
  //           <title>Lorem ipsum dolor</title>
  //         </head>
  //         <body>
  //           <p>
  //             Lorem ipsum dolor
  //           </p>
  //         </body>
  //       </html>
  //     `;
  //     replaceAll([
  //       {
  //         active: true,
  //         identifier: "1234",
  //         queries: ["ipsum"],
  //         queryPatterns: [],
  //         replacement: "sit",
  //       },
  //     ]);
  //     expect(document.documentElement).toMatchSnapshot();
  //   });
  //   it("does not overwrite 'script' elements", () => {
  //     document.documentElement.innerHTML = `
  //       <html>
  //         <head>
  //           <title>Lorem ipsum dolor</title>
  //           <script>
  //             function ipsum() {
  //             }
  //           </script>
  //         </head>
  //         <body>
  //           <p>
  //             Lorem ipsum dolor
  //           </p>
  //           <script>
  //             const html = document.createElement("html");
  //             html.innerHTML = "<body>Lorem ipsum dolor sit</body>";
  //           </script>
  //         </body>
  //       </html>
  //     `;
  //     replaceAll([
  //       {
  //         active: true,
  //         identifier: "1234",
  //         queries: ["ipsum"],
  //         queryPatterns: [],
  //         replacement: "sit",
  //       },
  //     ]);
  //     expect(document.documentElement).toMatchSnapshot();
  //   });
  //   it("maintains text enhancement elements", () => {
  //     document.documentElement.innerHTML = `
  //       <html>
  //         <body>
  //           <p>
  //             Lorem <u>ipsum</u> dolor
  //           </p>
  //         </body>
  //       </html>
  //     `;
  //     replaceAll([
  //       {
  //         active: true,
  //         identifier: "1234",
  //         queries: ["ipsum"],
  //         queryPatterns: [],
  //         replacement: "sit",
  //       },
  //     ]);
  //     expect(document.documentElement).toMatchSnapshot();
  //   });
  //   it("does not match across element boundaries", () => {
  //     document.documentElement.innerHTML = `
  //       <html>
  //         <body>
  //           <p>
  //             Lorem <u>ipsum</u> dolor
  //           </p>
  //         </body>
  //       </html>
  //     `;
  //     replaceAll([
  //       {
  //         active: true,
  //         identifier: "1234",
  //         queries: ["ipsum dolor"],
  //         queryPatterns: [],
  //         replacement: "sit",
  //       },
  //     ]);
  //     expect(document.documentElement).toMatchSnapshot();
  //   });
  // });
});
