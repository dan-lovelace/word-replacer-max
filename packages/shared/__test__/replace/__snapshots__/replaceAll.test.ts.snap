// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replaceAll document cases does not match across element boundaries 1`] = `
<html>
  <head />
  <body>
    
            
    <p>
      
              Lorem 
      <u>
        ipsum
      </u>
       dolor
            
    </p>
    
          
        
      
  </body>
</html>
`;

exports[`replaceAll document cases does not overwrite 'script' elements 1`] = `
<html>
  <head>
    
            
    <title
      data-is-replaced="1704067200000"
    >
      Lorem sit dolor
    </title>
    
            
    <script>
      
              function ipsum() {
                
              }
            
    </script>
    
          
  </head>
  
          
  <body>
    
            
    <p
      data-is-replaced="1704067200000"
    >
      
              Lorem sit dolor
            
    </p>
    
            
    <script>
      
              const html = document.createElement("html");
              html.innerHTML = "&lt;body&gt;Lorem ipsum dolor sit&lt;/body&gt;";
            
    </script>
    
          
        
      
  </body>
</html>
`;

exports[`replaceAll document cases includes 'title' elements in the head 1`] = `
<html>
  <head>
    
            
    <title
      data-is-replaced="1704067200000"
    >
      Lorem sit dolor
    </title>
    
          
  </head>
  
          
  <body>
    
            
    <p
      data-is-replaced="1704067200000"
    >
      
              Lorem sit dolor
            
    </p>
    
          
        
      
  </body>
</html>
`;

exports[`replaceAll document cases maintains text enhancement elements 1`] = `
<html>
  <head />
  <body>
    
            
    <p>
      
              Lorem 
      <u
        data-is-replaced="1704067200000"
      >
        sit
      </u>
       dolor
            
    </p>
    
          
        
      
  </body>
</html>
`;
